docs/proposals/http-tunnel-expose-strategy.md

```diff
@@ -0,0 +1,137 @@
+# **Exposing control-plane services using SNI and HTTP tunneling**
+
+Author: Youssef Azrak (@youssefazrak), Iacopo Rozzo (@irozzo-1A)
+
+Status: Draft proposal.
+
+
+## **Goals**
+
+Currently the control plane services running on the seed clusters can be exposed to the worker nodes relying on the strategies described in [this][k8c_expose_strategies] document.
+
+The main goal of this proposal is to provide a new strategy that routes the traffic to the target service based on SNI and HTTP Tunneling.
+
+
+## **Motivation and Background**
+
+User clusters rely on kubernetes default service to access the API server. Moreover services of type externalName are not currently supported (see [this][k8s_service_issue] issue), thus we have to rely on an IP address.
+
+Depending on the expose strategy in Kubermatic, we can either set up an endpoint IP to be hitting a Load balancer or a seed cluster node directly. The latter is called “NodePort strategy” and brings limitations: Usually those IPs are not static and if a node fails, there is no guarantee that it will be replaced by a new one having the same IP. Hence breaking the communications between the user clusters and their control plane. 
+
+To overcome this limitation, we have two strategies that rely on load balancers. 
+
+The first one: “Global LoadBalancer” is using one entry point and is routing requests based on ports uniquely assigned in the nodeport range for each user cluster service (apiserver and openvpn). One downside of this approach is having a limited number of listeners on the cloud load balancer. In AWS, there is a limit of 50 listeners per ELB (ALB or NLB) which means 50 services. As we are using 2 services per user cluster this leads to a maximum of 25 user clusters per load balancer. Another downside is that each service consumes two node ports, which could lead to exhaustion of the node port ranges for large clusters (e.g. default nodeport range is 30000-32767 meaning ~700 user clusters). 
+
+This leads us to the second load balancing strategy: “One Load Balancer Per User Cluster”. The downside of this solution is cost related. The cost of the load balancers increases linearly with the number of user clusters. 
```

Comment:
I think that in many cases the cost of the LBs is dominated by the cost of the rest of the infrastructure especially VMs (e.g. if the number of nodes per cluster is in the order of ~100). But we do have use cases in which a lot of small user cluster are created (e.g. ~5 nodes per user cluster).

Another problem, more relevant in general maybe, is related to LB quota issues (e.g. 50 by default in AWS): https://docs.aws.amazon.com/elasticloadbalancing/latest/application/load-balancer-limits.html